# Imports

In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from imblearn.over_sampling import RandomOverSampler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

---

# Global variables and constants

In [2]:
# Set the random state
rs=1

# Set the target names
target_names = ["Healthy Loan", "High-risk Loan"]

---

# Function Definitions

In [3]:
# Instantiates a Logistic Regression model and fits the model 
def logic_reg_model(X, y):
    """
    Create and train a Logistic Regression model.

   Input:
        X (pd.Dataframe): The feature data for training the model.
        y (pd.Series): The target labels for training the model.

    Output:
        sklearn.linear_model.LogisticRegression: A trained Logistic Regression model.
         
    """
    # Instantiate the Logistic Regression model
    model = LogisticRegression(random_state=rs)
    # Fit the model 
    model.fit(X, y)
    return model

In [4]:
# Counts the distinct values of the label
def count_distinct_values(series):
    """
    Count the distinct values in a Pandas Series and display the results.

    Input:
        series (array): The data to count distinct values.

    Output:
        value_counts (array): The distinct values and their counts.
        
    """
    # Count the distinct values
    value_counts = series.value_counts()
    
    # Display the distinct values
    print("Distinct values:")
    print(value_counts)
    
    #Return the count
    return value_counts

In [5]:
# Calculate the balanced accuracy score 
def balanced_accuracy(y_true, y_pred):
    """
    Calculate and return the balanced accuracy score for a classification model.

    Input:
        y_true (array): True target values.
        y_pred (array): Predicted target values from a classification model.

    Output:
        balanced_accuracy (float): The balanced accuracy score.

    """
    # Calculate the balanced accuracy score of the model
    balanced_accuracy = balanced_accuracy_score(y_true, y_pred)
    
    # Return the balanced accuracy score
    return balanced_accuracy

In [6]:
#  Generate a confusion matrix
def conf_matrix(y_true, y_pred, model_name):
    """
    Generate and display a confusion matrix for a classification model.

    Input:
        y_true (array): True target values.
        y_pred (array): Predicted target values from a classification model.
        model_name (string): Name or identifier of the classification model.

    Output:
        confusion (array): The confusion matrix as a 2D array.

    """
    # Generate the confusion matrix
    confusion = confusion_matrix(y_true, y_pred)
    
    # Print the result
    print(f"Confusion Matrix for {model_name}:")
    
    # Retutn the confusion matrix
    return confusion

In [7]:
# Generate a classification report
def class_report(y_true, y_pred, model_name, target_names):
    """
    Generate and display a classification report for a classification model.

    Input:
        y_true (array): True target values.
        y_pred (array): Predicted target values from a classification model.
        model_name (string): Name or identifier of the classification model.
        target_names (list): List of class labels.

    Output:
    Print the Classification Report for Logistic Regression
    
    """
    
    # Generate a classification report
    report = classification_report(y_true, y_pred, target_names=target_names)
    
    # Display the result
    print(f"Classification Report for {model_name}:")
    print(report)

---

# Main Code

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [8]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
data = Path('Resources/lending_data.csv')
df_lending = pd.read_csv(data)

# Review the DataFrame (First 5 rows and last 5 rows)
display(df_lending.head())
display(df_lending.tail())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  loan_status  
77531               12                 2       56600            1  
77532               11                 2       50900            1  
77533               11                 2       50300            1  
77534               10                 2       45300            1  
77535                9                 2       42300            1

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [9]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df_lending["loan_status"]                                                                                                               
# Separate the X variable, the features
X = df_lending.drop(columns="loan_status")

In [10]:
# Review the y variable Series (First 5 rows and last 5 rows)
print(y.head())
print(y.tail())

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64
77531    1
77532    1
77533    1
77534    1
77535    1
Name: loan_status, dtype: int64


In [11]:
# Review the X variable DataFrame (First 5 rows and last 5 rows)
display(X.head())
display(X.tail())

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

loan_size  interest_rate  borrower_income  debt_to_income  \
77531    19100.0         11.261            86600        0.653580   
77532    17700.0         10.662            80900        0.629172   
77533    17600.0         10.595            80300        0.626401   
77534    16300.0         10.068            75300        0.601594   
77535    15600.0          9.742            72300        0.585062   

       num_of_accounts  derogatory_marks  total_debt  
77531               12                 2       56600  
77532               11                 2       50900  
77533               11                 2       50300  
77534               10                 2       45300  
77535                9                 2       42300

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [12]:
# Check the balance of our target values
label_counts = count_distinct_values(y)

Distinct values:
0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [13]:
# Split the data using train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    random_state=rs,
                                                    stratify=y 
                                                    )
X_train.shape

(58152, 7)

There are 58152 data points in the training set.
There are 7 features in the training set.

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [14]:
# Instantiate the Logistic Regression model and Fit the model using training data
model_training= logic_reg_model(X_train, y_train)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [15]:
# Make a prediction using the testing data
y_pred_train = model_training.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [16]:
# Print the balanced accuracy score of the model
balanced_accuracy(y_test, y_pred_train)

0.9442676901753825

In [17]:
# Generate a confusion matrix for the model
conf_matrix(y_test, y_pred_train,"Logistic Regression Model with Training Data")


Confusion Matrix for Logistic Regression Model with Training Data:


array([[18679,    80],
       [   67,   558]], dtype=int64)

In [18]:
# Print the classification report for the model
class_report(y_test,y_pred_train, "Logistic Regression Model with Training Data",target_names=target_names)

Classification Report for Logistic Regression Model with Training Data:
                precision    recall  f1-score   support

  Healthy Loan       1.00      1.00      1.00     18759
High-risk Loan       0.87      0.89      0.88       625

      accuracy                           0.99     19384
     macro avg       0.94      0.94      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model with training data, predicts both the 0 (Healthy Loan) and 1 (High-risk Loan) labels effectively. It achieves a high precision of 1.00 for the Healthy Loan class, indicating its perfect performance to make accurate positive predictions, and a precision of 0.87 for the High-risk Loan class, indicating its effectiveness to accurately predict high-risk loans. This model achieves strong predictive performance in maintaining accuracy between the Healthy Loan and High-risk Loan classes, and strong performance in identifying both cases (more for negative cases )

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [19]:
# Instantiate the random oversampler model
ros = RandomOverSampler(random_state=rs)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)
len(X_resampled)

112554

In [20]:
# Count the distinct values of the resampled labels data
resampled_label_counts = count_distinct_values(y_resampled)

Distinct values:
0    56277
1    56277
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [21]:
# Instantiate the Logistic Regression model and Fit the model using the resampled training data
model_resampled= logic_reg_model(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred_resampled = model_resampled.predict(X_test)

### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [22]:
# Print the balanced_accuracy score of the model 

balanced_accuracy(y_test, y_pred_resampled)

0.9959744975744975

In [23]:
# Generate a confusion matrix for the model
conf_matrix(y_test, y_pred_resampled, "Logistic Regression Model with Resampled Data")

Confusion Matrix for Logistic Regression Model with Resampled Data:


array([[18668,    91],
       [    2,   623]], dtype=int64)

In [24]:
# Print the classification report for the model
class_report(y_test, y_pred_resampled, "Logistic Regression Model with Resampled Data", target_names)

Classification Report for Logistic Regression Model with Resampled Data:
                precision    recall  f1-score   support

  Healthy Loan       1.00      1.00      1.00     18759
High-risk Loan       0.87      1.00      0.93       625

      accuracy                           1.00     19384
     macro avg       0.94      1.00      0.96     19384
  weighted avg       1.00      1.00      1.00     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** The logistic regression model with oversampled data, predicts both the 0 (Healthy Loan) and 1 (High-risk Loan) labels effectively. It achieves a high precision of 1.00 for the Healthy Loan class, indicating its perfect performance to make accurate positive predictions, and a precision of 0.87 for the High-risk Loan class, indicating its effectiveness to accurately predict high-risk loans. This model achieves higher accuracy, a more balanced confusion matrix, and improved classification report metrics compared to the model trained on the original data.